### PR TITLE
feat: support certificate related operation

### PR DESCRIPTION
**Pull request by oil-oil**

### CODE REVIEW

File: certificate.go [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+package cloud
+
+import (
+	"context"
+	"path"
+	"time"
+)
+
+// Certificate is the definition of API7 Cloud Certificate.
+// some management fields.
```
Comment (tokers) [inline certificate.go:24]: Is the whole sentence smooth?

File: certificate.go [hunk 2]
```diff
@@ -0,0 +1,163 @@
...
+// CertificateSpec is the specification of the Certificate
+type CertificateSpec struct {
+	// The certificate in PEM format.
+	Certificate string `json:"certificate"`
+	// Private key is the private key for the corresponding certificate in PEM format.
+	PrivateKey string `json:"private_key"`
+	// CACertificate is the client ca certificate in PEM format used to verify client certificate.
+	CACertificate string `json:"ca_certificate,omitempty"`
+	// Labels are used for resource classification and indexing
+	Labels []string `json:"labels,omitempty"`
```
Comment (tokers) [inline certificate.go:49]: Add comment for `Type`.

Also, show optional values for Type in the comment.

File: certificate.go [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+	Type   string   `json:"type,omitempty"`
+}
+
+// CertificateInterface is the interface for manipulating Certificates.
+type CertificateInterface interface {
+	// CreateCertificate creates an API7 Cloud Certificate in the specified control plane.
+	// The given `cert` parameter should specify the desired Certificate specification.
+	// Users need to specify the ControlPlane in the `opts`.
+	// The returned Certificate will contain the same Certificate specification plus some
+	// management fields and default values.
```
Comment (tokers) [inline certificate.go:59]: Also tell users that the `PrivateKey` field will be empty in the returned object.

Comment (tokers) [reply certificate.go:59]: Same to `Update`, `Get` and `List`.

File: common_types.go [hunk 4]
```diff
@@ -153,6 +153,9 @@ type ResourceCreateOptions struct {
 	// in the Application. e.g., when creating API, CanaryRelease, the
 	// Application.ID should be specified.
 	Application *Application
+	// Certificate indicates which Certificate should this resource belong.
+	// This field should be specified when users want to update sub-resources
+	Certificate *Certificate
```
Comment (tokers) [inline common_types.go:158]: Why add the `Certificate` in these options?

Where do you use it?

File: testdata/test.pem [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+5Y9yzws7ZYtlaTZF9dljOIQrCpiYYOPUCuuCtQDFs1q8RIf2l9BLaHgGXpPont1X
+5c0vr0WTPgUQYsSbzMWEdn92k7a6+1SoCphyIiIj2OMV11avic4ODgSxG80Ia+q8
+NYyM2eaJchHmxku3wmf3YJqnsvq9NNDgPpN/y/fIyNnkl7YUQRMbKxZic5dHxp7O
+DgxeFG5090FGuWflsW+t2XHYCgr/GBoclKo/XW0fz+cSh5sl/YsgelQ3bp06c/uU
+E8Ex6QIDAQABo2IwYDAdBgNVHSUEFjAUBggrBgEFBQcDAQYIKwYBBQUHAwIwDAYD
+VR0TAQH/BAIwADAfBgNVHSMEGDAWgBSccme0YnV4aA/q6/pXsOmdoaGJXzAQBgNV
+HREECTAHggV2YXVsdDAKBggqhkjOPQQDAgNJADBGAiEAmukbKpOlu2szqG6AMLip
+lVUyicINprURYDQ4DyVE1N8CIQDTalKhDSmEyWALbTCqijCfptj1m8YdfL9OtLtk
+wt+74Q==
+-----END CERTIFICATE-----
```
Comment (tokers) [inline testdata/test.pem:18]: Missing the EOL.

File: certificate.go (hunk 1 above)
Comment (oil-oil) [reply certificate.go:24]: fixed

File: certificate.go (hunk 2 above)
Comment (oil-oil) [reply certificate.go:49]: > Add comment for `Type`.
> 
> Also, show optional values for Type in the comment.

Added comments.

File: certificate.go (hunk 3 above)
Comment (oil-oil) [reply certificate.go]: > Same to `Update`, `Get` and `List`.

Added comments.

File: common_types.go (hunk 4 above)
Comment (oil-oil) [reply common_types.go]: > Why add the `Certificate` in these options?
> 
> Where do you use it?

This field has been removed, I misunderstood it before.

File: testdata/test.pem (hunk 5 above)
Comment (oil-oil) [reply testdata/test.pem]: > Missing the EOL.

fixed.

File: certificate.go (hunk 2 above)
Comment (tokers) [reply certificate.go:49]: We don't use magic number of literal strings. Please refer to other constants.

File: certificate_test.go [hunk 6]
```diff
@@ -0,0 +1,317 @@
...
+		mockFunc      func(t *testing.T) httpClient
+	}{
+		{
+			name: "create successfully",
+			pendingCert: &Certificate{
+				CertificateSpec: CertificateSpec{
+					Certificate: string(cert),
+					PrivateKey:  string(key),
+					Labels:      nil,
+					Type:        "Client",
```
Comment (LiteSun) [inline certificate_test.go:53]: using constants

File: certificate_test.go [hunk 7]
```diff
@@ -0,0 +1,317 @@
...
+			expectedError: "",
+		},
+		{
+			name: "mock error",
+			pendingCert: &Certificate{
+				CertificateSpec: CertificateSpec{
+					Certificate: "invalid cert",
+					PrivateKey:  "invalid  key",
+					Labels:      nil,
+					Type:        "Client",
```
Comment (LiteSun) [inline certificate_test.go:72]: ditto

File: certificate_test.go [hunk 8]
```diff
@@ -0,0 +1,317 @@
...
+		pendingCert   *Certificate
+		expectedError string
+		mockFunc      func(t *testing.T) httpClient
+	}{
+		{
+			name: "update successfully",
+			pendingCert: &Certificate{
+				CertificateSpec: CertificateSpec{
+					Certificate: "invalid cert",
+					PrivateKey:  "invalid  key",
```
Comment (Baoyuantop) [inline certificate_test.go:118]: use correct cert?

Comment (oil-oil) [reply certificate_test.go]: > use correct cert?

fixed.

File: certificate_test.go (hunk 7 above)
Comment (oil-oil) [reply certificate_test.go]: > ditto

fixed.

File: certificate_test.go (hunk 6 above)
Comment (oil-oil) [reply certificate_test.go]: > using constants

fixed.

File: certificate.go (hunk 2 above)
Comment (oil-oil) [reply certificate.go:59]: > 

I draw the type as a constant.

File: certificate.go [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+type CertificateSpec struct {
+	// The certificate in PEM format.
+	Certificate string `json:"certificate"`
+	// Private key is the private key for the corresponding certificate in PEM format.
+	PrivateKey string `json:"private_key"`
+	// CACertificate is the client ca certificate in PEM format used to verify client certificate.
+	CACertificate string `json:"ca_certificate,omitempty"`
+	// Labels are used for resource classification and indexing.
+	Labels []string `json:"labels,omitempty"`
+	// Type is the certificate type, it can be "server" or "client", the default is "server".
```
Comment (tokers) [inline certificate.go:60]:
```suggestion
	// Type is the certificate type, optional values can be:
  	//   * ServerCertificate, server-type certificate.
  	//   * ClientCertificate, client-type certificate.
```

Comment (oil-oil) [reply certificate.go]: > 

fixed.

**Review by tokers — APPROVED** [review]

**Review by Baoyuantop — APPROVED** [review]

**Review by LiteSun — APPROVED** [review]

**Pull request merged**